src/main/java/com/facebook/openwifirrm/ucentral/UCentralClient.java

```diff
@@ -195,27 +202,125 @@ public boolean login() {
 		}
 
 		// Parse access token from response
-		JSONObject respBody;
+		WebTokenResult token;
 		try {
-			respBody = new JSONObject(response.getBody());
-		} catch (JSONException e) {
+			token = gson.fromJson(response.getBody(), WebTokenResult.class);
+		} catch (JsonSyntaxException e) {
 			logger.error("Login failed: Unexpected response", e);
 			logger.debug("Response body: {}", response.getBody());
 			return false;
 		}
-		if (!respBody.has("access_token")) {
+		if (token == null || token.access_token == null || token.access_token.isEmpty()) {
 			logger.error("Login failed: Missing access token");
-			logger.debug("Response body: {}", respBody.toString());
+			logger.debug("Response body: {}", response.toString());
 			return false;
 		}
-		this.accessToken = respBody.getString("access_token");
+		this.accessToken = token;
 		logger.info("Login successful as user: {}", username);
-		logger.debug("Access token: {}", accessToken);
+		logger.debug("Access token: {}", accessToken.access_token);
+		logger.debug("Refresh token: {}", accessToken.refresh_token);
 
 		// Load system endpoints
 		return loadSystemEndpoints();
 	}
 
+	/** 
+	 * when use public endpoints, refresh the access token if it's expired. 
+	 */
+	public synchronized void refreshAccessToken(){
+		if(usePublicEndpoints){
+			accessToken = getAccessToken();
+		}
+	}
+
+	/**
+	 * Check if an access token is expired.
+	 *
+	 * @return true if the access token is expired
+	 */
+	private boolean isAccessTokenExpired() {
+		if(accessToken == null){
+			return true;
+		}
+		return accessToken.created + TOKEN_REFRESH_WINDOW_S <
```

Comment:
You are right. There are two conception. One is access_token is expired and we need to refresh it. Another is that the token is completely expired even for a refresh request. (I call it a refresh token expired)

In the first version of this PR. There are  two methods, one is `isAccessTokenExpired`, which uses `idle_timeout` to check if an access token is expired, another is `isRefreshTokenExpired`, which uses `TOKEN_REFRESH_WINDOW_S ` to check if this token is completely expired.

The second version(the current version) of this PR. it only has `isAccessTokenExpired` and uses `TOKEN_REFRESH_WINDOW_S ` to check if an access token is expired. `access_token == null` to check if the access token is completely expired or not.

I guess you votes for `idle_timeout` and `expired_in` to check these two conditions?  Then config `TOKEN_REFRESH_WINDOW_S` seems not meaningful. I might also have some misunderstanding about the task description. What's the role of `TOKEN_REFRESH_WINDOW_S`? @elludraon 

> if owsec says a token is expired (e.g. through expires_in) and we consider it not expired then that would be a security concern. 

BTW, `expired_in` is a number which is far larger than `TOKEN_REFRESH_WINDOW_S`. So all the operations we have are in the scope of security I think.

Reply:
Yes I think we should use both - I actually think `idle_timeout` is what Jeffrey meant to have.

So the flow would look like this:

- Is there a token?
  - No: Login
  - Yes: go to next step
- is the token expired? (i.e. `now >= token.created + token.expires_in`)
  - No: Go to next step
  - Yes: login
- does the token need to be refreshed? (i.e. `now > last_access + token.idle_timeout`)
  - No: just use it!
  - Yes: refresh it

Reply:
Can you check if "idle_timeout" is actually implemented/enforced? I originally overlooked that field but I'm also not sure it's really used. If it is *not* used, then the purpose of `TOKEN_REFRESH_WINDOW_S` was some arbitrary time to refresh a token before it did expire (so that we can use the "refresh" flow at all, instead of completely re-login).

Reply:
I went through the code last week and it's not used atm (refreshing a token isn't a thing across the services right now) but I imagine it WILL be used and if so, we should follow it. Maybe good to clarify with someone regarding that.

Reply:
Then it seems that there is not need to configure `TOKEN_REFRESH_WINDOW_S`

Reply:
Ah, I see the problem with the original task. I did not notice that, when refreshed, the new token contains the *same* `created` time so that makes `TOKEN_REFRESH_WINDOW_S` useless. So we only need logic to handle:
- `idle_timeout` (shouldn't really happen, but might as well since the code is already done)
- relogin flow upon expiry

Reply:
Update: It appears that the "refreshToken" call is returning "created" time of the LAST token, not the current/refreshed one. So that is the confusion here. 😢